src/todos/todos.change-stream.database.service.ts

```diff
@@ -0,0 +1,87 @@
+import { Injectable } from '@nestjs/common'
+import { InjectModel } from '@nestjs/mongoose'
+import { Model } from 'mongoose'
+import type { ChangeStreamDocument } from 'mongodb'
+
+import { OPERATIONS } from '../common/constants/common.constants'
+import { TODO_MODEL, USER_MODEL } from '../common/constants/database.constants'
+import { ChangeStreamService } from '../common/services/change-stream.service'
+import type {
+  OperationCallback,
+  OperationType,
+} from '../common/types/common.types'
+import type { User } from '../user/schemas/user.schema'
+import type { Todo } from './schemas/todos.schema'
+
+@Injectable()
+export class TodosChangeStreamDatabaseService extends ChangeStreamService<Todo> {
+  eventListeners = new Map<OperationType, OperationCallback[]>()
+
+  constructor(
+    @InjectModel(TODO_MODEL) private readonly todoModel: Model<Todo>,
+    @InjectModel(USER_MODEL) private readonly userModel: Model<User>,
+  ) {
+    super(todoModel, [
+      {
+        $match: {
+          operationType: {
+            $in: [OPERATIONS.INSERT, OPERATIONS.UPDATE, OPERATIONS.DELETE],
+          },
+        },
+      },
+    ])
+  }
+
+  addEventListener(event: OperationType, callback: OperationCallback) {
+    if (this.eventListeners.has(event)) {
+      this.eventListeners.get(event)!.push(callback)
+    } else this.eventListeners.set(event, [callback])
+  }
+
+  protected async handleChange(changeStreamDoc: ChangeStreamDocument) {
```

Comment:
Please divide in three separate functions per event. This will lead to creation of three separate changestreams as well.